cwr/index.go

```diff
@@ -71,55 +87,259 @@ func (i *Indexer) Index(ctx context.Context, stream chan *cid.Cid) error {
+				if meta.IsPathNotFound(err) {
+					continue
+				} else if err != nil {
+					results <- err
+					return
+				}
+				numberOfTx := len(v.([]interface{}))
+
+				for j := 0; j < numberOfTx; j++ {
+					v, err := graph.Get("Groups", strconv.Itoa(k), "Transactions", field, strconv.Itoa(j))
+					if meta.IsPathNotFound(err) {
+						continue
+					} else if err != nil {
+						results <- err
+						return
+					}
+					tx, ok := v.(map[string]interface{})
+					if !ok {
+						results <- err
+						return
+					}
+					jobs <- jobIn{cwr.Cid(), tx, indexFn}
+				}
+			}
 		}
-		if err := i.indexPublisherControlledBySubmiter(cwrRecord.Cid().String(), publisherControlledBySubmitter); err != nil {
-			return err
+		close(jobs)
+	}()
+
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
+
+	for result := range results {
+		if result != nil {
+			err = result
+			continue //continue to drain results channel.
 		}
 	}
+	if err != nil {
+		return err
+	}
+	return i.sqlTx.Commit()
+}
 
-	return nil
+func (i *Indexer) worker(jobs <-chan jobIn, results chan<- error, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for job := range jobs {
+		results <- job.indexFn(job.cwrID, job.tx)
+	}
+}
+
+// indexRecord indexes a particular CWR record using the provided index
+// function.
+func (i *Indexer) indexRecord(cwrID, cid *cid.Cid, indexFn func(*cid.Cid, *meta.Object) error) error {
+	obj, err := i.store.Get(cid)
+	if err != nil {
+		return err
+	}
+	return indexFn(cwrID, obj)
 }
 
 // indexRegisteredWork indexes the given registeredWork record on its title,iswc,CompositeType and record_type
 // properties.
-func (i *Indexer) indexRegisteredWork(cid string, registeredWork *RegisteredWork) error {
-	log.Info("indexing registered work", "title", registeredWork.Title, "iswc", registeredWork.ISWC, "CompositeType", registeredWork.CompositeType, "Record Type", registeredWork.RecordType)
-	_, err := i.indexDB.Exec(
-		`INSERT INTO registered_work (object_id, title, iswc, composite_type,record_type) VALUES ($1, $2, $3, $4, $5)`,
-		cid, registeredWork.Title, registeredWork.ISWC, registeredWork.CompositeType, registeredWork.RecordType,
-	)
+func (i *Indexer) indexRegisteredWork(cwrID *cid.Cid, obj *meta.Object) error {
+	registeredWork := &RegisteredWork{}
+
+	if err := obj.Decode(registeredWork); err != nil {
+		return err
+	}
+
+	log.Info("indexing nwr (registered work)", "object_id", obj.Cid().String(), "Title", registeredWork.Title, "ISWC", registeredWork.ISWC, "Composite Type", registeredWork.CompositeType, "Record Type", registeredWork.RecordType)
+	stmt, err := i.sqlTx.Prepare(`INSERT INTO registered_work (cwr_id,object_id, title, iswc, composite_type,record_type) VALUES ($1, $2, $3, $4, $5, $6)`)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(cwrID.String(), obj.Cid().String(), registeredWork.Title, registeredWork.ISWC, registeredWork.CompositeType, registeredWork.RecordType)
```

Comment:
Why not just use [`tx.Exec`](https://golang.org/pkg/database/sql/#Tx.Exec)?

Reply:
you mean without the tx  prepare?

Reply:
Yes.

Reply:
It seems like using prepare might be more efficiant . https://en.wikipedia.org/wiki/Prepared_statement

Reply:
But you are preparing and then executing the statement once every time, so there are no performance benefits.